plugins/pluginmanagers/lz-n.nix

```diff
@@ -0,0 +1,106 @@
+          type = str;
+          description = "The plugin's name (not the module name). This is what is passed to the load(name) function.";
+        };
+        enabledInSpec = helpers.mkNullOrStrLuaFnOr bool ''
+          When false, or if the function returns false, then this plugin will not be included in the spec.
+          This option corresponds to the `enabled` property of lz.n.
+        '';
+        beforeAll = helpers.mkNullOrLuaFn "Always executed before any plugins are loaded.";
+        before = helpers.mkNullOrLuaFn "Executed before this plugin is loaded.";
+        after = helpers.mkNullOrLuaFn "Executed after this plugin is loaded.";
+        event = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on event. Events can be specified as BufEnter or with a pattern like BufEnter *.lua";
+        cmd = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on command.";
+        ft = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on filetype.";
+        colorscheme = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on colorscheme.";
+        keys = helpers.mkNullOrOption (listOf helpers.keymaps.mapOptionSubmodule) "Lazy-load on key mapping. Use the same format as `config.keymaps`.";
+        priority = helpers.mkNullOrOption number ''
+          Only useful for start plugins (not lazy-loaded) to force loading certain plugins first. 
+          Default priority is 50 (or 1000 if colorscheme is set).
+        '';
+        load = helpers.mkNullOrLuaFn "Can be used to override the `vim.g.lz_n.load()` function for this plugin.";
+      };
+    };
+in
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "lz-n";
+  originalName = "lz.n";
+  maintainers = [ helpers.maintainers.psfloyd ];
+  defaultPackage = pkgs.vimPlugins.lz-n;
+
+  settingsDescription = ''
+    Options provided to `vim.g.lz_n`.
+
+    `{ load = "fun"; }` -> `vim.g.lz_n = { load = fun, }`
+  '';
+
+  settingsOptions = {
+    load = helpers.defaultNullOpts.mkLuaFn "vim.cmd.packadd" ''
+      Function used by `lz.n` to load plugins.
+    '';
+  };
+
+  callSetup = false; # Does not use setup
+
+  extraOptions = {
+    plugins = mkOption {
+      description = "List of plugins processed by lz.n";
+      default = [ ];
+      type = types.listOf lzPluginType;
+    };
+  };
+
+  extraConfig =
+    cfg:
+    let
+      processKeymaps =
+        keymaps:
+        if keymaps == null then
+          null
+        else
+          map (
+            keymap:
+            {
+              __unkeyed_1 = keymap.key;
+              __unkeyed_2 = keymap.action;
+              inherit (keymap) mode;
+            }
+            // keymap.options
+          ) keymaps;
+      processPlugin =
+        plugin:
+        removeAttrs plugin [
+          "name"
+          "enabledInSpec"
+        ]
+        // {
+          __unkeyed = plugin.name;
+          enabled = plugin.enabledInSpec;
+          keys = processKeymaps plugin.keys;
+        };
```

Comment:
I'm not a big fan of having this processing step; it de-couples our options from what upstream actually expects.

I appreciate the mixed list/attrs table format is awkward to represent in nix, so I understand why you've done it this way. If this were a `settings` option though, I'd be saying "this doesn't comply with rfc42".

The per-plugin options added in #1875 could have its own schema that doesn't relate to lz.n's, because that will be _our_ options that we map internally to whatever plugin is implementing lazy-loading.

Here, however, we are attempting to represent a plugin's (lz.n's) upstream options, so I think it is best to follow rfc42.

@traxys @GaetanLepage what's your view?

Reply:
When removing magic for `enabledInSpec` and `keys`, the only thing that is left is to set the name as unkeyed.

Reply:
I'd prefer to avoid having any processing that changes the shape of arguments passed to plugins, it's confusing when people look at examples that are in lua initially.
Though our documentation & tooling on unkeyed tables may be lacking, but we should strive to fix that if it's the case

Reply:
I often mull over different ways to represent mixed list-attrs tables in nix, there's pros/cons to various approaches but none that map _super_ intuitively.

Something like this may be nice, but I don't think it's worth the effort migrating away from `__unkeyed-N`:

```nix
{
  _type = "lua-table";
  items = [ ];
  named = { };
}
```

---

Either way, while it is a little ugly, I agree we should avoid the processing step and let users define exactly what upstream expects via our option.

Generally speaking, having users come across `__raw` and `__unkeyed` in documentation and examples is good for exposure, so they can later figure out how to translate lua into nix.

For this plugin specifically, it's likely users won't have to interact with it directly, since #1875 will be more user-facing.